sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotUseCollectionInItsOwnMethodCalls.cs

```diff
@@ -0,0 +1,212 @@
+        private const string AlwaysEmptyCollectionMessage = "This operation always produces an empty collection.";
+        private const string AlwaysSameCollectionMessage = "This operation always produces the same collection.";
+        private const string AlwaysTrueMessage = "Comparing to itself always returns true.";
+        private const string AlwaysFalseMessage = "Comparing to itself always returns false.";
+        private const string UnexpectedBehaviorMessage = "This operation will probably result in an unexpected behavior.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<string> trackedMethodNames = ImmutableHashSet.Create("AddRange", "Concat",
+            "Except", "ExceptWith", "Intersect", "IntersectWith", "IsSubsetOf", "IsSupersetOf", "IsProperSubsetOf",
+            "IsProperSupersetOf", "Overlaps", "SequenceEqual", "SetEquals", "SymmetricExceptWith", "Union", "UnionWith");
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var invocation = (InvocationExpressionSyntax)c.Node;
+                    var invocationExpressionString = invocation.Expression.ToString();
+
+                    if (!trackedMethodNames.Any(method =>
+                            invocationExpressionString.EndsWith(method, StringComparison.Ordinal)))
+                    {
+                        return;
+                    }
+
+                    var operands = GetOperandsToCheckIfTrackedMethod(invocation, c.SemanticModel);
+                    if (operands != null &&
+                        EquivalenceChecker.AreEquivalent(operands.Left, operands.Right))
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, operands.Left.GetLocation(),
+                            additionalLocations: new[] { operands.Right.GetLocation() },
+                            messageArgs: new[] { operands.Right.ToString(), operands.ErrorMessage }));
+                    }
+                }, SyntaxKind.InvocationExpression);
+        }
+
+        private static OperandsToCheck GetOperandsToCheckIfTrackedMethod(InvocationExpressionSyntax invocation,
+            SemanticModel model)
+        {
+            var methodSymbol = model.GetSymbolInfo(invocation).Symbol as IMethodSymbol;
+
+            bool isEnumerableMethod;
+            var message = ProcessIssueMessageFromMethod(methodSymbol, out isEnumerableMethod);
+            if (message == null)
+            {
+                return null;
+            }
+
+            if (isEnumerableMethod && methodSymbol.MethodKind == MethodKind.Ordinary)
```

Comment:
Why do we need the isEnumerableMethod? To know how many arguments are in the expression? Perhaps we could use different variable name...

Actually, this could be a nice SE rule - currently you cannot recognize cases where one of the arguments is a variable assigned to the same instance...